File: messagematch/src/main/java/org/forwoods/messagematch/match/fieldmatchers/InstantTypeMatcher.java
package org.forwoods.messagematch.match.fieldmatchers;

import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.function.Function;

public class InstantTypeMatcher extends FieldMatcher<Instant> {

	private static Function<String, Instant> parses = Instant::parse;
	
	public InstantTypeMatcher(String binding, boolean nullable, FieldComparatorMatcher comparator) {
		super(binding, nullable, comparator);
	}

	@Override
	protected Instant asComparable(String val) {
		try {
			return parses.apply(val);
		} catch (DateTimeParseException e) {
			long millis = Long.parseLong(val);
			return Instant.ofEpochMilli(millis);
		}
	}

	@Override
	boolean doMatch(String value) {
		try {
			return parses.apply(value)!=null;
		} catch (DateTimeParseException e) {
			return false;
		}
	}

	@Override
	public boolean doSymRange(Instant value, Instant compareTo, String s) {
		return TimeTypeMatcher.doSymRangeTemporal(value, compareTo, s);
	}

	@Override
	public boolean doASymRange(Instant value, Instant compareTo, String s) {
		return TimeTypeMatcher.doASymRangeTemporal(value, compareTo, s);
	}


	//public so it can be overridden for custom date parsers
	public static void setParses(Function<String, Instant> parses) {
		InstantTypeMatcher.parses = parses;
	}
}

File: messagematch-typescript/src/match/InstantTypeMatcher.ts
import {FieldMatcher} from './fieldmatchers';
import {Temporal} from "@js-temporal/polyfill";
import TimeTypeMatcher from "./TimeTypeMatcher";

type T = Temporal.Instant;
// Port of Java InstantTypeMatcher. Uses Date objects for comparables.
export class InstantTypeMatcher extends FieldMatcher<T> {
  // parses: string -> Date

  constructor(binding: string | null, nullable: boolean, comparator: any) {
    super(binding, nullable, comparator);
  }

  public asComparable(val: string): T {
    try {
      return Temporal.Instant.from(val);
    }
    catch (e) {
      // if parse has failed try intepreting as epoch millis
      const ms = Number(val);
      return Temporal.Instant.fromEpochMilliseconds(ms);
    }
  }

  protected doMatch(value: string): boolean {
    try {
      return !!this.asComparable(value);
    }
    catch (e) {
      return false;
    }
  }

  // symmetric range: absolute difference <= range
  doSymRange( value: T, compareTo: T, s: string) : boolean {
    const actual: Temporal.Duration  = value.until(compareTo).abs();
    const range: Temporal.Duration = TimeTypeMatcher.getDuration(s).abs();
    return Temporal.Duration.compare(actual, range)<=0;
  }


  doASymRange(value: T,
                             compareTo: T,
                             s: string) : boolean {
    let start: T;
    let end: T;
    const range: Temporal.Duration = TimeTypeMatcher.getDuration(s);
    if (range.sign<0) {
      start = compareTo.add(range)
      end= compareTo;
    }
    else {
      start = compareTo;
      end = compareTo.add(range);
    }
    return Temporal.Instant.compare(value, start) >= 0 && Temporal.Instant.compare(value, end) <= 0;
  }

  compare(value: T, compareTo: T): 1 | 0 | -1 {
    return Temporal.Instant.compare(value, compareTo);
  }
}

export default InstantTypeMatcher;
